pytheus/mmap.py

```diff
@@ -0,0 +1,344 @@
+                self._padding = b"\0" * (BYTE_ALIGNMENT - remainder)
+        else:
+            self._aligned_size = self._size
+            if remainder != 0:
+                self._aligned_size += BYTE_ALIGNMENT
+                self._padding = b"\0" * (BYTE_ALIGNMENT - remainder)
+
+    def get_(self, position: int) -> StructFormatType:
+        return struct.unpack(self._struct_format_str, self._mmap[position:position+self._size])[0]
+
+    def set_(self, position: int, value: StructFormatType) -> None:
+        self._mmap[position:position+self._aligned_size] = \
+            struct.pack(self._struct_format_str, value) + self._padding
+
+    def size_(self, value: StructFormatType) -> int:
+        return self._aligned_size
+
+
+class StrMmapDictTypeBackend(MmapDictTypeBackend):
+    def __init__(
+        self, mmap: mmap.mmap,
+        encoding: str,
+    ) -> None:
+        super().__init__(mmap)
+        self._encoding = encoding
+
+    def get_(self, position: int) -> str:
+        offset_position = position + 4
+        return self._mmap[offset_position:offset_position+self._size_by_position(position)].decode(
+            self._encoding
+        )
+
+    def set_(self, position: int, value: str) -> None:  # type: ignore[override]
+        size = self._size_by_value(value)
+        padding = b""
+        remainder = size % BYTE_ALIGNMENT
+        if remainder != 0:
+            padding = b"\0" * (BYTE_ALIGNMENT - remainder)
+        offset_position = position + 4
+        self._mmap[position:offset_position] = struct.pack(STRUCT_INT_FORMAT, size)
+        self._mmap[offset_position:offset_position+size+len(padding)] = \
+            value.encode(self._encoding) + padding
+
+    def size_(self, value: str) -> int:  # type: ignore[override]
+        size_by_value = self._size_by_value(value)
+        combined_size = 4 + size_by_value
+        if combined_size < BYTE_ALIGNMENT:
+            return BYTE_ALIGNMENT
+        else:
+            remainder = combined_size % BYTE_ALIGNMENT
+            if remainder:
+                return combined_size + BYTE_ALIGNMENT - remainder
+            else:
+                return combined_size
+
+    def _size_by_position(self, position: int) -> int:
+        return struct.unpack(STRUCT_INT_FORMAT, self._mmap[position:position+4])[0]
+
+    def _size_by_value(self, value: str) -> int:
+        return len(value.encode(self._encoding))
+
+
+class MmapDict(collections.abc.MutableMapping):
+    """ Python dictionary that streams its data to a memory-mapped file.
+
+    ############
+    # Features #
+    ############
+
+    - It only support basic types: int, float, bool and string.
+
+    - It is not thread-safe for writing, which means only a single process should be writing.
+
+    - Index access will work in these two cases:
+
+        - In the write process, provided there is only one.
+
+        - In any read-only process, provided there is no write process running at the same time.
```

Comment:
This is ok for our use case. 

Only the worker process can write to its process database file (mmap file) in an indexed way, while the collection functionality will keep reading them all, with no need to index such reads as it needs all data. When the worker process dies, then it will aggregate its metrics in a history database file (mmap file), which can only be read/written by one process at a time (Will use a lock for this).